fix(ngModel): remove reference to parentForm from removed control

Narretz opened the pull request:
Otherwise, once the removed control's validity changes,
it will continue to be reflected on its former parent form.

Fixes #12263

@gkalpak can you take a look?

Narretz committed:
fix(ngModel): remove reference to parentForm from removed control

This fixes cases where the control gets removed, but the control's
element stays in the DOM.
Previously, if the removed control's validity changed, a reference to
it would again be stored on its former parent form.

Fixes #12263

gkalpak commented on src/ng/directive/form.js:
```diff
@@ -168,6 +168,7 @@ function FormController(element, attrs, $scope, $animate, $interpolate) {
     });
 
     arrayRemove(controls, control);
+    control.$$parentForm = nullFormCtrl;
```
I think a similar change is need in [`$addControl()`][1]: `control.$$parentForm = form`

[1]: https://github.com/angular/angular.js/pull/12418/files#diff-1838ed9bd28307b6d42de98549826ae5R134

caitp replied:
seems right

gkalpak commented on src/ng/directive/ngModel.js:
```diff
@@ -1311,7 +1309,7 @@ function addSetValidityMethod(context) {
 
     toggleValidationCss(validationErrorKey, combinedState);
-    parentForm.$setValidity(validationErrorKey, combinedState, ctrl);
+    ctrl.$$parentForm.$setValidity(validationErrorKey, combinedState, ctrl);
```
I think a similar change is needed in [`$setDirty()`][1]: `ctrl.$$parentForm.$setDirty()`

[1]: https://github.com/Narretz/angular.js/blob/fix-parent-form/src/ng/directive/form.js#L222

gkalpak commented:
Since nested `ngForm`s are also reated as controls to parent-forms (i.e. can be added/removed via `$add/removeControl()`), similar changes need to be made in `formDirective` (i.e. referencing `form.$$parentForm` instead of the _private_ `parentForm` variable.

Note, that this issue does not only affect validity states, but other properties or forms/controls, such as `$dirty`, `$submitted` etc.

It would also be nice to have more thorough testing (testing those other properties - not only `$error`).

--
But other than the comments above, it LGTM :smile:

gkalpak commented:
(And of course, there is the fail on Travis, but it's an easy fix :stuck_out_tongue:)

Narretz committed:
refactor(form, ngModel): streamline how controls are added to parent forms

This delegates setting the control's parentForm to the parentForm's
$addControl method. This way, the model controller saves one instance
of looking up the parentForm controller. The form controller keeps two
lookups (one for its own ctrl, one for the optional parent).

This also fixes adding the parentForm in the following case:
- a control is removed from a parent, but its corresponding DOM
element is not destroyed
- the control is then re-added to the form

Before the fix, the control's parentForm was only set once during
controller initialization, so the the parentForm would not be set on
the control in that specific case.

Narretz commented:
@gkalpak Thanks for the review! Do you remember what failure Travis recorded? It's such a pain to look through the logs.

caitp commented:
> @gkalpak Thanks for the review! Do you remember what failure Travis recorded? It's such a pain to look through the logs.

Running "jshint:ng" (jshint) task
   src/ng/directive/ngModel.js
    241 |  var parentForm = this.$$parentForm = $element.inheritedData('$formController') || nullFormCtrl
                                                                                                         ^ Missing semicolon.
>> 1 error in 70 files
Warning: Task "jshint:ng" failed. Use --force to continue.
Aborted due to warnings.
The command "./scripts/travis/build.sh" exited with 3.
cache.2
store build cache
1.74snothing changed, not updating cache
$ ./scripts/travis/print_logs.sh

Narretz committed:
fix(form, ngModel): correctly notify parent form when children are added

Test that re-added controls propagate validity changes to the parent form.

Ensure that when a form / control that was removed and then attached
to a different parent, is renamed / deleted, the new parent will
be notified of the change.

Document that dynamic adding / removing of controls may require manually
propagating the current state of the control to the parent form.

Narretz commented on test/ng/directive/ngModelSpec.js:
```diff
@@ -19,7 +19,6 @@ describe('ngModel', function() {
       };
 
       element = jqLite('<form><input></form>');
-      element.data('$formController', parentFormCtrl);
```
Since the parentFormCtrl is now extracted via regular require, this mechanism doesn't work anymore. This might break people's tests.

Narretz commented:
Thanks @caitp 
This turned out to be a bit more complex, since the adding of controls is handled differently in forms / ngModel. I refactored this now, so that the initial adding of controls happens in the preLink function for both form and ngModel (this actually saves one instance of accessing the parent form controller in ngModel). The logic of assigning the parentForm to the control is moved to the $addControl.
When a control is added, its validity, dirty and submit states will be reflected automatically on the parent form. 
I'll have to test if these changes impact performance.

Narretz commented:
@gkalpak / @caitp can you please take another look?

caitp commented on src/ng/directive/form.js:
```diff
@@ -131,6 +128,28 @@ function FormController(element, attrs, $scope, $animate, $interpolate) {
+
+    control.$$parentForm = form;
+
+    forEach(control.$pending, function(value, name) {
```
this may not be a big deal in most cases, but 3x `forEach()` calls seems like a perf hit, which would be most noticeable in forms with visibility controlled by ng-if or ng-switch. I'm not saying it's a huge problem, but yeah

caitp commented:
apart from 1 comment, 3rd commit looks ok to me

gkalpak commented on src/ng/directive/form.js:
```diff
@@ -131,6 +128,28 @@ function FormController(element, attrs, $scope, $animate, $interpolate) {
+    }
+
+    if (control.$submitted && !form.$submitted) {
+      form.$setSubmitted();
```
Do we really want this ?
It is really a corner case, but if we already have unsubmitted form (with more fields/child-forms), then adding a submitted child-form shouldn't necessarily mark the form as submitted.
(The same holds for a submitted form and unsubmitted child-form being added.)

Neither way seems intuitive (so maybe it's better to leave the property untouched).

gkalpak replied:
I guess this is in accordance to what happens in `$setSubmitted()` (where the parent form is also marked as submitted), but it feels unintuitive.

Anyway, since this is how things are, let's leave it as is.

gkalpak commented on test/ng/directive/formSpec.js:
```diff
@@ -58,6 +58,78 @@ describe('form', function() {
+    form.$removeControl(form.control);
+    expect(form.control).toBeUndefined();
+    expect(form.$error.maxlength).toBeFalsy();
+    expect(form.$dirty).toBe(false);
```
I would move these 3 expectations at line 119 (after the input has been edited, but before it has been added to the form), in order to verify that the form has been affected **only** after the control has been added back).
(It's not a big deal, since the previous test already verifies that a removed control does not affect the form, but it still makes me feel better :smiley:)

gkalpak commented on test/ng/directive/formSpec.js:
```diff
@@ -577,6 +649,83 @@ describe('form', function() {
+    });
+
+
+    it('should react ti changes in manually added child forms', function() {
```
_ti_  -->  _to_

gkalpak commented on test/ng/directive/formSpec.js:
```diff
@@ -577,6 +649,83 @@ describe('form', function() {
+        '</form>')(scope);
+
+      var form = scope.myForm;
+      var childformController = doc.find('ng-form').eq(0).controller('form');
```
You know you could simply write `form.childform`, right ? :stuck_out_tongue: 
<sub>(It applies in all tests.)</sub>

Narretz replied:
I think I was trying to be safe in case the form gets removed etc. But I guess it does work with these tests

gkalpak commented on src/ng/directive/form.js:
```diff
@@ -490,7 +513,9 @@ var formDirectiveFactory = function(isNgForm) {
 
-            var parentFormCtrl = controller.$$parentForm;
+            var parentFormCtrl = ctrls[1] || ctrls[0].$$parentForm;
+            parentFormCtrl.$addControl(controller);
```
Hard-coding `parentFormCtrl`suffers from the same issues in [`attr.$observe(...)`][1] and [`formElement.on('$destroy', ...)`][2]; i.e. if the form has been since removed and re-added to another parent form, it won't be properly renamed/removed.

You need to use `ctrls[0].$$parentForm`.

(It's OK to use `parentFormCtrl` on line 517 though.)

[1]: https://github.com/angular/angular.js/pull/12418/files#diff-1838ed9bd28307b6d42de98549826ae5R523
[2]: https://github.com/angular/angular.js/pull/12418/files#diff-1838ed9bd28307b6d42de98549826ae5R531

gkalpak commented on src/ng/directive/form.js:
```diff
@@ -204,7 +224,7 @@ function FormController(element, attrs, $scope, $animate, $interpolate) {
       }
     },
-    parentForm: parentForm,
+    parentForm: form.$$parentForm,
```
This property is not used, since you are accessing the parent form via `ctrl.$$parentForm` inside `addSetValidity()`.

gkalpak commented on src/ng/directive/ngModel.js:
```diff
@@ -1022,7 +1020,7 @@ var ngModelDirective = ['$rootScope', function($rootScope) {
         pre: function ngModelPreLink(scope, element, attr, ctrls) {
           var modelCtrl = ctrls[0],
-              formCtrl = ctrls[1] || nullFormCtrl;
+              formCtrl = ctrls[1] || ctrls[0].$$parentForm;
```
Similar problem as https://github.com/angular/angular.js/pull/12418#discussion_r38092011.

Narretz committed:
style(formSpec): fix indentation

Narretz commented:
Thanks for the review @gkalpak, I have fixed the cases where the wrong parent was referenced.
I've also reverted the behavior where calling addControl would try to set the current state (error, success, pending, dirty, submitted) on the parentForm. This way, we match the previous behavior, and we avoid possible performance penalties. I've added documentation about this.

gkalpak commented on test/ng/directive/formSpec.js:
```diff
@@ -58,6 +58,114 @@ describe('form', function() {
+    changeInputValue(input, 'a');
+
+    expect(form.$error.maxlength).toBeFalsy();
+  });
```
In this test you are not actually testing that the form has reacted to changes in the added control.
After you add the control you only check `form.$error.maxlength` (which hasn't changed since the last time you checked).

gkalpak commented on test/ng/directive/formSpec.js:
```diff
@@ -58,6 +58,114 @@ describe('form', function() {
+    scope.$digest();
+
+    expect(form.childControlMoved).toBeUndefined();
+    expect(otherForm.childControlMoved).toBeUndefined();
```
lol

Narretz replied:
Hmm I don't understand!

gkalpak replied:
Sorry - I was just thinking loud (laughing actually).

I find it funny (in a good way), that you can have a control be one form's child (in DOM), but belong to another form (in Angular terms).
I thought it was the wrong expectation at first site, but it is correct and find it nice that it works.

Sorry for the noise and confusion :disappointed_relieved:

gkalpak commented on test/ng/directive/formSpec.js:
```diff
@@ -547,35 +655,147 @@ describe('form', function() {
+      form.$addControl(childformController);
+      changeInputValue(input, 'a');
+
+      expect(form.$error.maxlength).toBeFalsy();
```
Same as https://github.com/angular/angular.js/pull/12418/files#r38523628.

gkalpak commented:
Apart for a couple of nitpicks in test, it LGTM :+1:

Narretz commented:
Thx again for the reviews! Landed as f8a07dd^..aa11dfc
